File: attractor/agent/steering.py
"""Steering and follow-up message queues."""

from __future__ import annotations

import asyncio

from ..llm.models import Message


class SteeringManager:
    """Manages steering and follow-up message injection."""

    def __init__(self) -> None:
        self._steering: asyncio.Queue[str] = asyncio.Queue()
        self._followups: asyncio.Queue[str] = asyncio.Queue()

    def steer(self, message: str) -> None:
        """Add a message to inject between tool rounds."""
        self._steering.put_nowait(message)

    def follow_up(self, message: str) -> None:
        """Add a message to process after current turn completes."""
        self._followups.put_nowait(message)

    def drain_steering(self) -> list[Message]:
        """Get all pending steering messages as user Messages."""
        messages: list[Message] = []
        while not self._steering.empty():
            try:
                text = self._steering.get_nowait()
                messages.append(Message.user(text))
            except asyncio.QueueEmpty:
                break
        return messages

    def drain_followups(self) -> list[Message]:
        """Get all pending follow-up messages as user Messages."""
        messages: list[Message] = []
        while not self._followups.empty():
            try:
                text = self._followups.get_nowait()
                messages.append(Message.user(text))
            except asyncio.QueueEmpty:
                break
        return messages

File: internal/agent/steering.go
package agent

import (
	"sync"

	"github.com/nigelpepper/attractor/internal/llm"
)

// SteeringManager holds steering and follow-up messages injected between rounds.
type SteeringManager struct {
	mu        sync.Mutex
	steering  []string
	followups []string
}

// NewSteeringManager returns an empty manager.
func NewSteeringManager() *SteeringManager { return &SteeringManager{} }

// Steer queues a message to inject between tool rounds.
func (s *SteeringManager) Steer(message string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.steering = append(s.steering, message)
}

// FollowUp queues a message to process after the current turn.
func (s *SteeringManager) FollowUp(message string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.followups = append(s.followups, message)
}

// DrainSteering returns and clears pending steering messages.
func (s *SteeringManager) DrainSteering() []llm.Message {
	s.mu.Lock()
	defer s.mu.Unlock()
	msgs := make([]llm.Message, 0, len(s.steering))
	for _, t := range s.steering {
		msgs = append(msgs, llm.UserMessage(t))
	}
	s.steering = nil
	return msgs
}

// DrainFollowups returns and clears pending follow-up messages.
func (s *SteeringManager) DrainFollowups() []llm.Message {
	s.mu.Lock()
	defer s.mu.Unlock()
	msgs := make([]llm.Message, 0, len(s.followups))
	for _, t := range s.followups {
		msgs = append(msgs, llm.UserMessage(t))
	}
	s.followups = nil
	return msgs
}
